src/Core/Managed/Shared/Extensibility/MetricAggregatorManager.cs

```diff
@@ -0,0 +1,275 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Globalization;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+#if CORE_PCL || NET45 || NET46
+    using TaskEx = System.Threading.Tasks.Task;
+#endif
+
+    /// <summary>
+    /// Represents a hub for metric aggregation procedures.
+    /// </summary>
+    public sealed class MetricAggregatorManager : IDisposable
+    {
+        /// <summary>
+        /// Name of the property added to aggregation results to indicate duration of the aggregation interval.
+        /// </summary>
+        private static string intervalDurationPropertyName = "IntervalDurationMs";
+
+        /// <summary>
+        /// Reporting frequency.
+        /// </summary>
+        private static TimeSpan snapshotFrequency = TimeSpan.FromMinutes(1);
+
+        /// <summary>
+        /// Last time snapshot was initiated.
+        /// </summary>
+        private DateTimeOffset lastSnapshotStartDateTime;
+
+        /// <summary>
+        /// Cancellation token source to allow cancellation of the snapshotting task.
+        /// </summary>
+        private CancellationTokenSource cancellationSource;
+
+        /// <summary>
+        /// Metric aggregator snapshotting task.
+        /// </summary>
+        private Task snapshotTask;
+
+        /// <summary>
+        /// A dictionary of all metric aggregators instantiated via this manager.
+        /// </summary>
+        private ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator> aggregatorDictionary;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricAggregatorManager"/> class.
+        /// </summary>
+        public MetricAggregatorManager()
+            : this(new TelemetryClient())
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MetricAggregatorManager"/> class.
+        /// </summary>
+        /// <param name="client">Telemetry client to use to output aggregated metric data.</param>
+        public MetricAggregatorManager(TelemetryClient client)
+        {
+            this.Client = client ?? new TelemetryClient();
+            this.aggregatorDictionary = new ConcurrentDictionary<MetricAggregator, SimpleMetricStatisticsAggregator>();
+
+            this.lastSnapshotStartDateTime = DateTimeOffset.UtcNow;
+
+            this.cancellationSource = new CancellationTokenSource();
+
+            this.snapshotTask = new Task(this.SnapshotRunner, TaskCreationOptions.LongRunning);
```

Comment:
It makes the third thread SDK runs in background. QP configuration, Performance Counters collection and this aggregator. Can you think of any options how to combine this thread with Performance Counters collection thread?